File: sdks/sdk-go/sigilum/rfc9421_conformance_test.go
package sigilum

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"testing"
)

type rfcFixture struct {
	Fixed struct {
		Created int64  `json:"created"`
		Nonce   string `json:"nonce"`
	} `json:"fixed"`
	Vectors []struct {
		Name                  string   `json:"name"`
		Method                string   `json:"method"`
		URL                   string   `json:"url"`
		Body                  *string  `json:"body"`
		ExpectedTargetURI     string   `json:"expected_target_uri"`
		ExpectedContentDigest string   `json:"expected_content_digest"`
		ExpectedComponents    []string `json:"expected_components"`
	} `json:"vectors"`
}

func loadRFCFixture(t *testing.T) rfcFixture {
	t.Helper()
	path := filepath.Join("..", "..", "shared", "test-vectors", "http-signatures-rfc9421.json")
	bytes, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fixture rfcFixture
	if err := json.Unmarshal(bytes, &fixture); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	return fixture
}

func TestRFC9421ProfileVectorsAndStrictChecks(t *testing.T) {
	fixture := loadRFCFixture(t)
	tmp := t.TempDir()
	if _, err := InitIdentity(InitIdentityOptions{Namespace: "alice", HomeDir: tmp}); err != nil {
		t.Fatalf("init identity: %v", err)
	}
	identity, err := LoadIdentity(LoadIdentityOptions{Namespace: "alice", HomeDir: tmp})
	if err != nil {
		t.Fatalf("load identity: %v", err)
	}

	for _, vector := range fixture.Vectors {
		var body []byte
		if vector.Body != nil {
			body = []byte(*vector.Body)
		}

		signed, err := SignHTTPRequest(identity, SignRequestInput{
			URL:     vector.URL,
			Method:  vector.Method,
			Body:    body,
			Created: fixture.Fixed.Created,
			Nonce:   fixture.Fixed.Nonce,
		})
		if err != nil {
			t.Fatalf("sign %s: %v", vector.Name, err)
		}

		if signed.URL != vector.ExpectedTargetURI {
			t.Fatalf("%s: unexpected URL: %s", vector.Name, signed.URL)
		}
		signatureInput := signed.Headers["signature-input"]
		if !strings.Contains(signatureInput, "created="+strconv.FormatInt(fixture.Fixed.Created, 10)) {
			t.Fatalf("%s: created missing in Signature-Input: %s", vector.Name, signatureInput)
		}
		if !strings.Contains(signatureInput, "nonce=\""+fixture.Fixed.Nonce+"\"") {
			t.Fatalf("%s: nonce missing in Signature-Input: %s", vector.Name, signatureInput)
		}
		expectedComponents := "(" + quoteComponents(vector.ExpectedComponents) + ")"
		if !strings.Contains(signatureInput, expectedComponents) {
			t.Fatalf("%s: components mismatch in Signature-Input: %s", vector.Name, signatureInput)
		}
		if vector.ExpectedContentDigest != "" && signed.Headers["content-digest"] != vector.ExpectedContentDigest {
			t.Fatalf("%s: digest mismatch", vector.Name)
		}

		seen := map[string]struct{}{}
		ok := VerifyHTTPSignature(VerifySignatureInput{
			URL:               signed.URL,
			Method:            signed.Method,
			Headers:           signed.Headers,
			Body:              signed.Body,
			ExpectedNamespace: "alice",
			NowUnix:           fixture.Fixed.Created + 5,
			MaxAgeSeconds:     60,
			SeenNonces:        seen,
		})
		if !ok.Valid {
			t.Fatalf("%s: expected valid strict verify, got: %s", vector.Name, ok.Reason)
		}

		replay := VerifyHTTPSignature(VerifySignatureInput{
			URL:               signed.URL,
			Method:            signed.Method,
			Headers:           signed.Headers,
			Body:              signed.Body,
			ExpectedNamespace: "alice",
			NowUnix:           fixture.Fixed.Created + 5,
			MaxAgeSeconds:     60,
			SeenNonces:        seen,
		})
		if replay.Valid || !strings.Contains(strings.ToLower(replay.Reason), "replay") {
			t.Fatalf("%s: expected replay detection, got: %#v", vector.Name, replay)
		}

		stale := VerifyHTTPSignature(VerifySignatureInput{
			URL:               signed.URL,
			Method:            signed.Method,
			Headers:           signed.Headers,
			Body:              signed.Body,
			ExpectedNamespace: "alice",
			NowUnix:           fixture.Fixed.Created + 500,
			MaxAgeSeconds:     60,
		})
		if stale.Valid || (!strings.Contains(strings.ToLower(stale.Reason), "expired") && !strings.Contains(strings.ToLower(stale.Reason), "valid")) {
			t.Fatalf("%s: expected stale signature rejection, got: %#v", vector.Name, stale)
		}
	}
}

func quoteComponents(components []string) string {
	quoted := make([]string, 0, len(components))
	for _, c := range components {
		quoted = append(quoted, "\""+c+"\"")
	}
	return strings.Join(quoted, " ")
}

File: sdks/sdk-ts/src/__tests__/rfc9421-conformance.test.ts
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import { initIdentity, loadIdentity } from "../identity-store.js";
import { signHttpRequest, verifyHttpSignature } from "../http-signatures.js";

type Vector = {
  name: string;
  method: string;
  url: string;
  body: string | null;
  expected_target_uri: string;
  expected_method_component: string;
  expected_content_digest?: string;
  expected_components: string[];
};

type NegativeVector = {
  name: string;
  source_vector: string;
  mutation: "method" | "header" | "body";
  field?: string;
  value: string;
  expected_reason_contains: string;
};

type Fixture = {
  fixed: { created: number; nonce: string };
  vectors: Vector[];
  negative_vectors?: NegativeVector[];
};

function readFixture(): Fixture {
  const fixturePath = path.resolve(
    process.cwd(),
    "../test-vectors/http-signatures-rfc9421.json",
  );
  return JSON.parse(fs.readFileSync(fixturePath, "utf8")) as Fixture;
}

describe("RFC 9421 profile conformance", () => {
  const tempDirs: string[] = [];

  afterEach(() => {
    while (tempDirs.length > 0) {
      const dir = tempDirs.pop();
      if (dir && fs.existsSync(dir)) {
        fs.rmSync(dir, { recursive: true, force: true });
      }
    }
  });

  function makeHomeDir(): string {
    const dir = fs.mkdtempSync(path.join(os.tmpdir(), "sigilum-sdk-"));
    tempDirs.push(dir);
    return dir;
  }

  it("matches shared vectors and enforces strict replay/timestamp", () => {
    const fixture = readFixture();
    const homeDir = makeHomeDir();
    initIdentity({ namespace: "alice", homeDir });
    const identity = loadIdentity({ namespace: "alice", homeDir });
    const signedByName = new Map<string, ReturnType<typeof signHttpRequest>>();

    for (const vector of fixture.vectors) {
      const signed = signHttpRequest(identity, {
        url: vector.url,
        method: vector.method,
        body: vector.body,
        created: fixture.fixed.created,
        nonce: fixture.fixed.nonce,
      });

      expect(signed.url).toBe(vector.expected_target_uri);
      expect(signed.method.toLowerCase()).toBe(vector.expected_method_component);
      signedByName.set(vector.name, signed);

      const signatureInput = signed.headers.get("signature-input");
      expect(signatureInput).toBeTruthy();
      expect(signatureInput).toContain(`created=${fixture.fixed.created}`);
      expect(signatureInput).toContain(`nonce="${fixture.fixed.nonce}"`);
      const expectedComponents = vector.expected_components
        .map((c) => `"${c}"`)
        .join(" ");
      expect(signatureInput).toContain(`(${expectedComponents})`);

      if (vector.expected_content_digest) {
        expect(signed.headers.get("content-digest")).toBe(vector.expected_content_digest);
      }

      const nonceStore = new Set<string>();
      const verifyOk = verifyHttpSignature({
        url: signed.url,
        method: signed.method,
        headers: signed.headers,
        body: signed.body ?? null,
        expectedNamespace: "alice",
        strict: {
          now: fixture.fixed.created + 5,
          maxAgeSeconds: 60,
          nonceStore,
        },
      });
      expect(verifyOk.valid).toBe(true);

      const replay = verifyHttpSignature({
        url: signed.url,
        method: signed.method,
        headers: signed.headers,
        body: signed.body ?? null,
        expectedNamespace: "alice",
        strict: {
          now: fixture.fixed.created + 5,
          maxAgeSeconds: 60,
          nonceStore,
        },
      });
      expect(replay.valid).toBe(false);
      expect(replay.reason).toMatch(/replay/i);

      const stale = verifyHttpSignature({
        url: signed.url,
        method: signed.method,
        headers: signed.headers,
        body: signed.body ?? null,
        expectedNamespace: "alice",
        strict: {
          now: fixture.fixed.created + 500,
          maxAgeSeconds: 60,
        },
      });
      expect(stale.valid).toBe(false);
      expect(stale.reason).toMatch(/expired|valid/i);
    }

    for (const negative of fixture.negative_vectors ?? []) {
      const baseSigned = signedByName.get(negative.source_vector);
      expect(baseSigned, `${negative.name}: source vector not found`).toBeTruthy();
      if (!baseSigned) {
        continue;
      }

      const headers = new Headers(baseSigned.headers);
      let method = baseSigned.method;
      let body = baseSigned.body ?? null;

      if (negative.mutation === "method") {
        method = negative.value;
      } else if (negative.mutation === "header") {
        const field = (negative.field ?? "").trim().toLowerCase();
        expect(field.length, `${negative.name}: missing header field`).toBeGreaterThan(0);
        headers.set(field, negative.value);
      } else if (negative.mutation === "body") {
        body = negative.value;
      }

      const result = verifyHttpSignature({
        url: baseSigned.url,
        method,
        headers,
        body,
        expectedNamespace: "alice",
        strict: {
          now: fixture.fixed.created + 5,
          maxAgeSeconds: 60,
        },
      });
      expect(result.valid, `${negative.name}: expected verify failure`).toBe(false);
      const normalizedReason = (result.reason?.toLowerCase() ?? "")
        .replace(/[^a-z0-9]+/g, " ")
        .trim();
      const normalizedExpected = negative.expected_reason_contains
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, " ")
        .trim();
      expect(normalizedReason).toContain(normalizedExpected);
    }
  });
});
